src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,3047 @@
+        BEAST_EXPECT(!passesLocalChecks(stx, reason));
+        BEAST_EXPECT(reason == "Cannot submit pseudo transactions.");
+        env.app().openLedger().modify([&](OpenView& view, beast::Journal j) {
+            auto const result = ripple::apply(env.app(), view, stx, tapNONE, j);
+            BEAST_EXPECT(!result.applied && result.ter == temINVALID_FLAG);
+            return result.applied;
+        });
+    }
+
+    void
+    testBatchWithSelfSubmit(FeatureBitset features)
+    {
+        testcase("batch with self submit");
+        // IMPORTANT: When a transaction is self submitted and another
+        // transaction is part of the batch, the batch might fail because the
+        // sequence is out of order. This is because the canonical order of
+        // transactions is determined by the account first. So in this case,
+        // alice's batch comes after bobs self submitted transaction even though
+        // the self submitted was after the batch.
+
+        using namespace test::jtx;
+        using namespace std::literals;
+
+        test::jtx::Env env{*this, envconfig()};
+        XRPAmount const baseFee = env.current()->fees().base;
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+
+        env.fund(XRP(1000), alice, bob);
+        env.close();
+
+        env(noop(bob), ter(tesSUCCESS));
+        env.close();
+
+        auto const aliceSeq = env.seq(alice);
+        auto const preAlice = env.balance(alice);
+        auto const preBob = env.balance(bob);
+        auto const bobSeq = env.seq(bob);
+
+        // Alice Pays Bob (Self Submit)
+        env(pay(alice, bob, XRP(10)), ter(tesSUCCESS));
+
+        // Alice & Bob Atomic Batch
+        auto const batchFee = batch::calcBatchFee(env, 1, 2);
+        env(batch::outer(alice, aliceSeq + 1, batchFee, tfAllOrNothing),
+            batch::inner(pay(alice, bob, XRP(10)), aliceSeq + 2),
+            batch::inner(pay(bob, alice, XRP(5)), bobSeq + 1),
+            batch::sig(bob));
+        auto const txIDs = env.tx()->getBatchTransactionIDs();
+        TxID const parentBatchId = env.tx()->getTransactionID();
+        std::vector<TestBatchData> testCases = {
+            {"tesSUCCESS", to_string(txIDs[0])},
+            {"tesSUCCESS", to_string(txIDs[1])},
+        };
+
+        // Bob pays Alice (Self Submit)
+        env(pay(bob, alice, XRP(5)), ter(tesSUCCESS));
+        env.close();
+
+        validateBatch(env, parentBatchId, testCases, 2);
+
+        // Alice consumes sequences (# of txns)
+        BEAST_EXPECT(env.seq(alice) == aliceSeq + 3);
+
+        // Alice consumes sequences (# of txns)
+        BEAST_EXPECT(env.seq(bob) == bobSeq + 2);
+
+        // Alice pays XRP & Fee; Bob receives XRP & pays Fee
+        BEAST_EXPECT(
+            env.balance(alice) == preAlice - XRP(10) - batchFee - baseFee);
+        BEAST_EXPECT(env.balance(bob) == preBob + XRP(10) - baseFee);
+    }
+
+    static std::unique_ptr<Config>
+    makeConfig(
```

Comment:
```suggestion
    makeSmallQueueConfig(
```
Name should be a bit more specific (and probably at the top of the file with the other helper functions)